nirtorch/from_nir.py

```diff
@@ -45,15 +47,31 @@ def execution_order_up_to_node(
         return execution_order + [node]
 
 
+@dataclasses.dataclass
+class GraphExecutorState:
+    """State for the GraphExecutor that keeps track of both the state of hidden
+    units and caches the output of previous modules, for use in (future) recurrent
+    computations."""
+
+    state: Dict[str, Any] = dataclasses.field(default_factory=dict)
+    cache: Dict[str, Any] = dataclasses.field(default_factory=dict)
+
+
 class GraphExecutor(nn.Module):
     def __init__(self, graph: Graph) -> None:
         super().__init__()
         self.graph = graph
+        self.stateful_modules = {}
         self.instantiate_modules()
         self.execution_order = self.get_execution_order()
         if len(self.execution_order) == 0:
             raise ValueError("Graph is empty")
 
+    def _is_module_stateful(self, module: torch.nn.Module) -> bool:
```

Comment:
The logic here implies that if any module has multiple inputs, it will be assumed to be stateful. This is a deal breaker!

Reply:
I agree, we need to find a better way to implement this.. It currently breaks in snnTorch because you may have multiple inputs but not be stateful (if the node keeps track of its own hidden state by itself)

Reply:
I'm happy to find other ways of doing this. But how?

Here's the challenge as far as I can tell
- Most frameworks can live without state (snnTorch, Sinabs, Rockpool)
- Norse requires a `state` parameter (similar to PyTorch RNNs)
- snnTorch can take `spk` and `mem` inputs

Would an option be to look for `state` in the arguments to account for the norse case and `spk` and `mem` to account for the snnTorch case?